### ISSUE TITLE
difference in behavior when returning a list of subtypes directly vs. with an intermediate variable

### ISSUE THREAD

**Issue by mskoh52** [microsoft/pyright#3513]
This code gives an error, since List is invariant:

```python
class A:
  pass

class B(A):
  pass

def test(x: List[A]) -> List[A]:
  blist = [B()]
  return blist
```

```
[Pyright reportGeneralTypeIssues] [E] Expression of type "list[B]" cannot be assigned to return type "List[A]"
  TypeVar "_T@list" is invariant
    "B" is incompatible with "A"
```

but if I change `test` to:

```python
def test(x: List[A]) -> List[A]:
  return [B()]
```

then no error is reported. It's weird because when I show the type of `blist` in the first example, it says `List[B]`, so I guess the type inference on that particular list is doing what I would have expected.

Similarly behavior on the function's argument:

![image](https://user-images.githubusercontent.com/11641396/170536057-54126e09-cb5a-4a97-9cb6-6f853df588eb.png)

where the error on that red highlighted word is 

```
[Pyright reportGeneralTypeIssues] [E] Argument of type "list[B]" cannot be assigned to parameter "x" of type "List[A]" in function "test"
  TypeVar "_T@list" is invariant
    "B" is incompatible with "A"
```

Apologies if this is already described in another issue. I searched a few terms but couldn't find anything exactly matching. I've also read the [type inference doc](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md) but that didn't really seem to answer this.

**Comment by JelleZijlstra** [comment]
This is related to bidirectional inference: https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types

**Comment by erictraut** [comment]
Yes, this is expected behavior. The inferred type of `[B()]` is normally `list[B]`. With the additional context in your second sample, bidirectional type inference produces the inferred type of `list[A]`. So this behavior is correct and expected.
